packages/flutter_tools/lib/src/crash_reporting.dart

```diff
@@ -0,0 +1,383 @@
+/// it must be supplied in the request.
+String _kStackTraceFilename = 'stacktrace_file';
+
+bool _testing = false;
+
+/// Sends crash reports to Google.
+class CrashReportSender {
+  static final Uri _defaultBaseUri = new Uri(
+      scheme: 'https',
+      host: _kCrashServerHost,
+      port: 443,
+      path: _kCrashEndpointPath,
+  );
+
+  static Uri _baseUri = _defaultBaseUri;
+  final Usage _usage = Usage.instance;
+
+  /// Sends one crash report.
+  ///
+  /// The report is populated from data in [error] and [stackTrace].
+  Future<Null> sendReport({@required dynamic error, @required dynamic stackTrace}) async {
+    try {
+      // TODO(yjbanov): we only actually send crash reports in tests. When we
+      // iron out the process, we will remove this guard and report crashes
+      // when !flutterUsage.suppressAnalytics.
+      if (!_testing) {
+        return null;
+      }
+
+      if (_usage.suppressAnalytics) {
+        return null;
+      }
+
+      printStatus('Sending crash report to Google.');
+
+      String version = FlutterVersion.getVersionString();
+
+      Uri uri = _baseUri.replace(
+        queryParameters: <String, String>{
+          'product': _kProductId,
+          'version': '$version',
+        },
+      );
+
+      _MultipartRequest req = new _MultipartRequest('POST', uri);
+      req.fields['product'] = _kProductId;
+      req.fields['version'] = '$version';
+      req.fields['type'] = _kDartTypeId;
+      req.fields['error_runtime_type'] = '${error.runtimeType}';
+
+      Chain chain = stackTrace is StackTrace
+          ? new Chain.forTrace(stackTrace)
+          : new Chain.parse(stackTrace.toString());
+
+      req.files.add(new http.MultipartFile.fromString(
+        _kStackTraceFileField,
+        '${chain.terse}',
+        filename: _kStackTraceFilename,
+      ));
+      http.StreamedResponse resp = await req.send();
+
+      if (resp.statusCode == 200) {
+        String reportId = await new http.ByteStream(resp.stream)
+            .bytesToString();
+        printStatus('Crash report sent (report ID: $reportId)');
+      } else {
+        printError('Failed to send crash report. Server responded with HTTP status code ${resp.statusCode}');
+      }
+    } catch(sendError, sendStackTrace) {
+      if (sendError is SocketException) {
+        printError('Failed to send crash report due to a network error: $sendError');
+      } else {
+        // If the sender itself crashes, just print. We did our best.
+        printError('Crash report sender itself crashed: $sendError\n$sendStackTrace');
+      }
+    }
+  }
+}
+
+/// Overrides [CrashReportSender._baseUri] for testing and enables testing
+/// mode.
+@visibleForTesting
+void overrideBaseCrashUrlForTesting(Uri uri) {
+  CrashReportSender._baseUri = uri;
+  _testing = true;
+}
+
+/// Resets [CrashReportSender._baseUri] back to the original value and disables
+/// test mode.
+@visibleForTesting
+void resetBaseCrashUrlForTesting() {
```

Comment:
I doubt we need to ever reset it (same probably applies to the other testing `reset*` methods...)